### PR TITLE
Introduce DotNetBuildRepo, DotNetBuild, and DotNetBuildPhase

### PR DESCRIPTION
**Pull request by mmitche**
Introduce DotNetBuildRepo, DotNetBuild, and DotNetBuildPhase
- DotNetBuildRepo (formerly DotNetBuildOuterRepo) is the marker for building the repo in .NET product mode
- DotNetBuild is calculated off any of DotNetBuildRepo, DOtNetBuildInnerRepo, and DotNetBuildOrchestrator being true
- Factored non-repo layout properties into a new file, RepoDefaults.props, instead of housing these in RepoLayout.props
- Used DotNetBuildPhase == 'foo' in some cases where it's simpler to do so.

### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/arcade/tree/main/Documentation/Validation

### CODE REVIEW

**Comment by mmitche** [comment]
https://github.com/dotnet/source-build/issues/3938

**Comment by mmitche** [comment]
https://github.com/dotnet/source-build/issues/3941

**Comment by mmitche** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=2352751&view=results official build

**Comment by mmitche** [comment]
This needs more verification. Not all the switches are totally active yet. For instance, the outer repo switch isn't passed from the orchestrator yet, neither is the orchestrator switch. I added some usages where I know they will be though.

File: eng/common/build.sh [hunk 1]
```diff
@@ -22,6 +22,9 @@ usage()
   echo "  --sourceBuild              Source-build the solution (short: -sb)"
   echo "                             Will additionally trigger the following actions: --restore, --build, --pack"
   echo "                             If --configuration is not set explicitly, will also set it to 'Release'"
+  echo "  --dotnetBuild              Build the solution in the way it will be built in the full .NET product (VMR) build (short: -db)"
+  echo "                             Will additionally trigger the following actions: --restore, --build, --pack"
+  echo "                             If --configuration is not set explicitly, will also set it to 'Release'"
```
Comment (ViktorHofer) [inline eng/common/build.sh:27]: Doesn't this also need to go into eng/common/build.ps1?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/RepoDefaults.props [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  <PropertyGroup>
+    <Configuration Condition="'$(Configuration)' == ''">Debug</Configuration>
+    <Platform Condition="'$(Platform)' == ''">AnyCPU</Platform>
+    <PlatformName Condition="'$(PlatformName)' == ''">$(Platform)</PlatformName>
+  </PropertyGroup>
+
+  <!-- DotNetBuild properties -->
+  <PropertyGroup>
+    <!-- The DotNetBuild property is set to true if in any kind of .NET product build (orchestrator, repo, or inner repo). -->
+    <DotNetBuild Condition="'$(DotNetBuildInnerRepo)' == 'true' or '$(DotNetBuildRepo)' == 'true'">Build</DotNetBuild>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/RepoDefaults.props:17]:
```suggestion
    <DotNetBuild Condition="'$(DotNetBuildInnerRepo)' == 'true' or '$(DotNetBuildRepo)' == 'true'">true</DotNetBuild>
```

File: eng/common/build.sh (hunk 1 above)
Comment (mmitche) [reply eng/common/build.sh:27]: Ya, will update.

**Comment by ViktorHofer** [comment]
> This needs more verification.

Yes, agreed. You could test this in a local VMR build and make the orchestrator pass the settings in.

**Comment by mmitche** [comment]
I'm doing some last minute testing but I think this is ready to go. Verified that the expected controls are showing up. Also enabled DotNetBuildRepo when we get into source build or vertical build modes, so we can transition off of the control use in those repos.

File: eng/common/build.sh [hunk 3]
```diff
@@ -22,6 +22,9 @@ usage()
   echo "  --sourceBuild              Source-build the solution (short: -sb)"
   echo "                             Will additionally trigger the following actions: --restore, --build, --pack"
   echo "                             If --configuration is not set explicitly, will also set it to 'Release'"
+  echo "  --dotnetBuild              Build the solution in the way it will be built in the full .NET product (VMR) build (short: -dnb)"
```
Comment (MichaelSimons) [inline eng/common/build.sh:25]: On first read this name felt a bit ambiguous to me.  Whenever you do any type of build on a repo that is part of the dotnet org you are doing a "dotnet build".  Would `productBuild` be more meaningful?

**Review by MichaelSimons — COMMENTED** [review]
Overall looks good.  I have one naming comment.

File: eng/common/build.sh (hunk 3 above)
Comment (mmitche) [reply eng/common/build.sh:25]: Ooh I like that much better. Are you suggesting also to change the other switch prefixes? (`DotNetBuild`) or leave those?

Comment (mmitche) [reply eng/common/build.sh:25]: Alright, all for that change.

**Review by MichaelSimons — APPROVED** [review]

File: eng/common/build.sh (hunk 3 above)
Comment (ViktorHofer) [reply eng/common/build.sh]: I don't like the inconsistency but I agree that the current switch's name isn't great.

**Comment by mmitche** [comment]
Verified in a full VMR build. Still needs some additional enabling there but looking good so far.

**Pull request merged**